options/rtld/generic/linker.cpp

```diff
@@ -896,6 +902,17 @@ void ObjectRepository::_parseDynamic(SharedObject *object) {
 		object->soName = reinterpret_cast<const char *>(object->baseAddress
 				+ object->stringTableOffset + *soname_offset);
 	}
+	if(object->haveTextRel) {
+		for(auto &phdr_ptr : object->exec_phdrs) {
+			elf_phdr* phdr = (elf_phdr *)phdr_ptr;
+			void* addr = (void *)(phdr->p_vaddr + (uintptr_t)object->baseAddress);
+			int prot = PROT_WRITE | PROT_READ | PROT_EXEC;
+			if (mlibc::sys_vm_protect)
+				mlibc::sys_vm_protect(addr, phdr->p_memsz, prot);
+			else
+				__ensure(!"sys_vm_protect required when DF_TEXTREL/DT_TEXTREL is present");
+		}
+	}
```

Comment:
This could be done in `linkObjects()` to be symmetrical with the code path that removes `PROT_WRITE` again.